### PR TITLE
Refactor register mov logic into genmovreg

### PR DESCRIPTION
**Pull request by SSoulaimane**

### CODE REVIEW

File: src/dmd/backend/cod3.d [hunk 1]
```diff
@@ -2595,21 +2595,74 @@ code *genmovreg(uint to,uint from)
...
+    // register kind. ex: GPR,XMM,SEG
+    static uint _K(uint reg)
+    {
+        switch (reg)
+        {
+        case ES:                   return ES;
+        case XMM15:
+        case XMM0: .. case XMM7:   return XMM0;
+        case AX:   .. case R15:    return AX;
+        default:                   return reg;
```
Comment (WalterBright) [inline src/dmd/backend/cod3.d]: ???? shouldn't this case be a compiler bug?

File: src/dmd/backend/cod3.d [hunk 2]
```diff
@@ -2595,21 +2595,74 @@ code *genmovreg(uint to,uint from)
...
     {
-        if (to == ES)
-            genregs(cdb,0x8E,0,from);
-        else if (from == ES)
-            genregs(cdb,0x8C,0,to);
-        else
-            genregs(cdb,0x89,from,to);
-        if (I64)
-            code_orrex(cdb.last(), REX_W);
+        if (tym == TYMAX) tym = TYsize_t; // avoid register slicing
```
Comment (WalterBright) [inline src/dmd/backend/cod3.d]: ??? register slicing ???

File: src/dmd/backend/cod3.d [hunk 3]
```diff
@@ -2595,21 +2595,74 @@ code *genmovreg(uint to,uint from)
...
+
+void genmovreg(ref CodeBuilder cdb, uint to, uint from, tym_t tym)
+{
+    // register kind. ex: GPR,XMM,SEG
+    static uint _K(uint reg)
+    {
+        switch (reg)
+        {
+        case ES:                   return ES;
+        case XMM15:
```
Comment (WalterBright) [inline src/dmd/backend/cod3.d]: ????

File: src/dmd/backend/cod3.d (hunk 1 above)
Comment (SSoulaimane) [reply src/dmd/backend/cod3.d]: There is an `assert(0)` in line `2661`.

File: src/dmd/backend/cod3.d (hunk 2 above)
Comment (SSoulaimane) [reply src/dmd/backend/cod3.d]: Line `2598` uses `TYMAX` as a placeholder type.  `TYMAX` has no size, so if no size specified, this moves a full register.

File: src/dmd/backend/cod3.d (hunk 3 above)
Comment (SSoulaimane) [reply src/dmd/backend/cod3.d]: https://github.com/dlang/dmd/blob/27097d85732c28954679153dba2965bfbbc37a34/src/dmd/backend/code_x86.d#L59-L61
https://github.com/dlang/dmd/blob/27097d85732c28954679153dba2965bfbbc37a34/src/dmd/backend/code_x86.d#L68
https://github.com/dlang/dmd/blob/27097d85732c28954679153dba2965bfbbc37a34/src/dmd/backend/code_x86.d#L76-L79
between `XMM7` and `XMM15` are non `XMM` registers.

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**